docs/design/2021-08-18-charsets.md

```diff
@@ -0,0 +1,271 @@
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Development Plan](#development-plan)
+
+# Introduction
+
+Currently, TiDB only supports ascii, binary, latin1, utf-8, and utf8mb4 character sets. This proposal aims to add a framework for character sets to TiDB to facilitate the support of other characters. This design takes the gbk character set as an example.
+
+# Motivation or Background
+
+Character: Various words and symbols are collectively referred to as characters. A character can be a Chinese character, an English letter, an Arabic numeral, a punctuation mark, a graphic symbol or a control symbol, etc.
+
+Charset: A collection of multiple characters is called a charset. Common character set names: ASCII character set, Unicode character set, GBK character set, etc.
+
+Character Encoding: Character encoding can be considered as a mapping rule. According to this mapping rule, a character can be mapped into other forms of data for storage and transmission in the computer. Each character set has corresponding character encoding rules, and the commonly used character set encoding rules include UTF-8 encoding, GBK encoding and so on.
+
+MySQL supports many character sets, such as `utf8mb4`, `gbk`, `gb18030`, etc. However, currently TiDB only supports 5 character sets, and there are still some problems with related character sets ([charset: incorrect encoding for `latin1` character set](https://github.com/pingcap/tidb/issues/18955)，[extended latin character set not support](https://github.com/pingcap/tidb/issues/3888)). Now it is difficult for TiDB to add a new character set. This proposal describes the implementation of the character set framework. In many traditional industries in China, `gbk`/`gb18030` is used, so this proposal uses gbk character set support as an example.
+
+## Characteristics
+
+Support gbk character set related functions:
+- Supports all code points defined by the gbk standard. Supports all code points defined by the gbk standard.
+- Support conversion with other character sets.
+  - Support data automatically converted according to the corresponding character set.
+  - Supports the use of `CAST`, `CONVERT` and other functions to convert gbk characters.
+- Support SQL statements, such as: `SET CHARACTER SET GBK`, `SET NAMES GBK`, `SHOW CHARSET`, etc.
+- It supports the comparison between strings in the gbk character set, as well as the comparison between gbk strings and strings in other character sets.
+- There is no mapping between (U+D800, U+DFFF) and gbk. Attempting to convert code points in this range will return "?".
```

Comment:
if it is talking about converting the surrogate code points from UTF8MB4 to GBK i don't know why they need to be singled out here. there are tons of real characters which is valid in UTF8MB4 but invalid in GBK e.g. 

```sql
select 
    column_0 utf8mb4, 
    convert(column_0 using gbk) gbk, 
    convert(column_0 using gb18030) gb18030
from (values row('🤔'), row('ς'), row('σ')) _;
```

```
+---------+------+---------+
| utf8mb4 | gbk  | gb18030 |
+---------+------+---------+
| 🤔      | ?    | 🤔     |
| ς       | ?    | ς       |
| σ       | σ    | σ       |
+---------+------+---------+
3 rows in set (0.00 sec)
```

Reply:
This is a description item referring to the gb18030 character set supported by MySQL:
https://dev.mysql.com/doc/refman/8.0/en/charset-gb18030.html#:~:text=There%20is%20no%20mapping%20between%20(U%2BD800%2C%20U%2BDFFF)%20and%20GB18030.%20Attempted%20conversion%20of%20code%20points%20in%20this%20range%20returns%20%27%3F%27.

Reply:
GB18030 is not GBK though

Reply:
Yes, I found GB18030 has special instructions for this: https://en.wikipedia.org/wiki/GB_18030#:~:text=0080%20%E2%80%93%20FFFF%20except%20D800%20%E2%80%93%20DFFF%5B.
Done.